### DISCUSSION TITLE
Obtaining host IP address when in docker

### DISCUSSION THREAD

**Discussion by larrywalters** [siderolabs/talos#6115] [Q&A] [answered]
Hello, I'm very new to Talos, so I'm not sure if this is documented somewhere, but I have been having trouble figuring it out. I am running the local docker quickstart via `talosctl cluster create --wait`. When that is ready, I `kubectl apply...` my resources. One of them is a Deployment that connects to a MySQL database. The database is running, for now, on my local machine, and thus outside the cluster.

I want to connect to this database from inside the Deployment Pod. My setup works with `minikube`, where I was able to obtain the IP address of my local "host" machine using `minikube ssh "cat /etc/hosts" | grep host.minikube.internal | cut -f 1`. However, since Talos doesn't have ssh, nor the specific variable that minikube has, I want to find something equivalent. But I have not been able to. I know docker has `host.docker.internal`, but that is from a container running simply inside docker. Here we have the containers for the whole `master` and `worker` nodes, and then in them we want the Pods to have access to this IP. And I could not find `host.docker.internal` existing in one of my Pods anyway.

Any help with this would be appreciated! Thanks.

**Comment by smira** [comment, marked as answer]
Talos doesn't do anything different in terms of Docker setup.

I believe the address you're looking for is the Docker bridge address, so it should be `10.5.0.1` with the default options of `talosctl cluster create`. 

Another way to see this address is with `docker network inspect talos-default`.

**Reply by smira** [reply]
A note: this should work on Linux. When using Docker on OS X/Windows, it might be different as Docker runs inside a VM, so routing to the host is much more tricky.

**Reply by larrywalters** [reply]
Thank you @smira for the answer! From my testing, on Mac/Windows, the gateway does appear initially as `10.5.0.1`, but then after cluster creation, `talosctl cluster show` indicates the gateway is `<nil>`. I'm not sure why. The way I worked around it is to connect the Talos docker containers to the Docker Desktop `bridge` network, and access the local machine through its gateway (see more details below). If anyone has a better solution, I'm open to hearing it!

**Comment by larrywalters** [comment]
@smira Thank you for the reply! I wanted to try some things before I responded.
I did attempt to use 10.5.0.1 as you suggested, but it did not work. Interestingly, I do see that this IP is listed as the gateway right after I start the cluster, but then afterwards the gateway is `<nil>`. Here is a simple example of just creating the cluster and then querying its status afterwards.
```
>$ talosctl cluster create --wait
validating CIDR and reserving IPs
generating PKI and tokens
creating network talos-default
creating master nodes
creating worker nodes
renamed talosconfig context "talos-default" -> "talos-default-58"
waiting for API
bootstrapping cluster
waiting for etcd to be healthy: OK
waiting for etcd members to be consistent across nodes: OK
waiting for apid to be ready: OK
waiting for kubelet to be healthy: OK
waiting for all nodes to finish boot sequence: OK
waiting for all k8s nodes to report: OK
waiting for all k8s nodes to report ready: OK
waiting for all control plane components to be ready: OK
waiting for kube-proxy to report ready: OK
waiting for coredns to report ready: OK
waiting for all k8s nodes to report schedulable: OK

merging kubeconfig into "/Users/lwalters/.kube/config"
renamed cluster "talos-default" -> "talos-default-52"
renamed auth info "admin@talos-default" -> "admin@talos-default-52"
renamed context "admin@talos-default" -> "admin@talos-default-52"
PROVISIONER       docker
NAME              talos-default
NETWORK NAME      talos-default
NETWORK CIDR      10.5.0.0/24
NETWORK GATEWAY   10.5.0.1
NETWORK MTU       1500

NODES:

NAME                      TYPE           IP         CPU    RAM      DISK
/talos-default-master-1   controlplane   10.5.0.2   2.00   2.1 GB   -
/talos-default-worker-1   worker         10.5.0.3   2.00   2.1 GB   -
>$ talosctl cluster show
PROVISIONER       docker
NAME              talos-default
NETWORK NAME      talos-default
NETWORK CIDR      10.5.0.0/24
NETWORK GATEWAY   <nil>
NETWORK MTU       1500

NODES:

NAME                     TYPE           IP         CPU   RAM   DISK
talos-default-master-1   controlplane   10.5.0.2   -     -     -
talos-default-worker-1   worker         10.5.0.3   -     -     -
```
When I try to inspect the `talos-default` network, I indeed do not see a gateway
```
>$ docker network inspect talos-default
[
    {
        "Name": "talos-default",
        ...
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.5.0.0/24"
                }
            ]
        },
        ...
    }
]

```
_Do you know why this would be happening?_

---
Now, I was able to get things working by using the gateway of docker's `bridge` network, which does show a gateway upon inspection.
```
>$ docker network inspect bridge
[
    {
        "Name": "bridge",
        ...
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": <...>,
                    "Gateway": <...>
                }
            ]
        },
        ...
    }
]

```
I connected the two talos containers to the `bridge` network, and am able to use its gateway to reach outside the cluster on my local machine.
```
>$ docker network connect bridge talos-default-master-1
>$ docker network connect bridge talos-default-worker-1
```
I am using macOS, and you did say things could be more tricky there. I also plan to get this working on Windows.
_If there is a detriment in doing it this way, please let me know._
Thanks!

**Reply by larrywalters** [reply]
@smira I unmarked your comment as the answer to see if you had any thoughts on my above question, continuing the discussion. If that's not proper protocol, my apologies, please let me know. Thanks.

**Reply by smira** [reply]
no problem, I have no clue about OS X, it's not really a Talos question, but rather Docker/OS X question, so I'm not that helpful. 

It will be great if you could provide a solution for those looking for an answer.

**Reply by larrywalters** [reply]
@smira Thanks for the response. I see the same behavior on Windows with Docker (running my same script to setup the cluster). The Gateway is at first 10.5.0.1, and then shows nil. So you're saying on Linux, that `NETWORK GATEWAY   10.5.0.1` will be persistent, and will not go to `<nil>` shortly after cluster creation, and that this is an issue probably with Docker on Mac/Windows?

**Reply by smira** [reply]
On Linux container addresses are directly routeable from the host, so the problem doesn't even exist )

**Reply by larrywalters** [reply]
@smira Ok, thank you for the information. I must admit, I'm mostly a Mac user. 🙂 The solution I've come up with so far is the one I mentioned above, regarding connecting the containers to Docker's `bridge` network. I will accept your original answer, since it is valid for Linux, and will make a reply above about what what I did on the other platforms. Thanks for the help!
